File: src/CVerbsBench/src/IbLib/connection.c
/**
 * @file connection.c
 * @author Fabian Ruhland, HHU
 * @date May 2018
 *
 * @brief Contains structs and functions to work with connections.
 */


#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <infiniband/verbs.h>
#include "connection.h"
#include "log.h"

void init_connection(connection* conn, ib_device *device, ib_prot_dom *prot_dom, ib_comp_queue *send_comp_queue,
                     ib_comp_queue *recv_comp_queue, ib_shared_recv_queue *recv_queue, uint64_t buf_size,
                     uint32_t queue_size) {
    LOG_INFO("CONNECTION", "Initializing connection...");

    conn->prot_dom = prot_dom;

    // Initialize queue pair
    conn->queue_pair = malloc(sizeof(ib_queue_pair));
    init_queue_pair(conn->queue_pair, prot_dom, send_comp_queue, recv_comp_queue, recv_queue, queue_size);

    conn->send_comp_queue = send_comp_queue;
    conn->recv_comp_queue = recv_comp_queue;

    // Initialize memory regions
    conn->send_mem_reg = malloc(sizeof(ib_mem_reg));
    init_mem_reg(conn->send_mem_reg, buf_size);

    conn->recv_mem_reg = malloc(sizeof(ib_mem_reg));
    init_mem_reg(conn->recv_mem_reg, buf_size);

    // Register memory regions
    register_memory_region(conn->prot_dom, conn->send_mem_reg);
    register_memory_region(conn->prot_dom, conn->recv_mem_reg);

    // Initialize scatter-gather elements
    conn->send_sge.addr = (uint64_t) conn->send_mem_reg->addr;
    conn->send_sge.length = (uint32_t) buf_size;
    conn->send_sge.lkey = conn->send_mem_reg->lkey;

    conn->recv_sge.addr = (uint64_t) conn->recv_mem_reg->addr;
    conn->recv_sge.length = (uint32_t) buf_size;
    conn->recv_sge.lkey = conn->recv_mem_reg->lkey;

    // Initialize work requests
    posix_memalign((void **) &conn->send_wrs, (size_t) getpagesize(), sizeof(struct ibv_send_wr) * queue_size);
    posix_memalign((void **) &conn->recv_wrs, (size_t) getpagesize(), sizeof(struct ibv_recv_wr) * queue_size);

    memset(conn->send_wrs, 0, sizeof(struct ibv_send_wr) * queue_size);
    memset(conn->recv_wrs, 0, sizeof(struct ibv_recv_wr) * queue_size);

    // Initialize the local connection information
    conn->local_conn_info.lid = device->lid;
    conn->local_conn_info.qpn = conn->queue_pair->qpn;
    conn->local_conn_info.rkey = conn->recv_mem_reg->rkey;
    conn->local_conn_info.remote_address = (uint64_t) conn->recv_mem_reg->addr;

    // Reset the remote connection information
    conn->remote_conn_info.lid = 0;
    conn->remote_conn_info.qpn = 0;
    conn->remote_conn_info.rkey = 0;
    conn->remote_conn_info.remote_address = 0;

    LOG_INFO("CONNECTION", "Finished initializing connection!");
}

void connect_to_server(connection *conn, const char *bind_address, const char *hostname, uint16_t port) {
    struct sockaddr_in server;
    struct hostent *he;
    char *ip_address;

    LOG_INFO("CONNECTION", "Connecting to server '%s'...", hostname);

    // Resolve hostname to an ip-address
    he = gethostbyname(hostname);

    if(he == NULL) {
        LOG_ERROR_AND_EXIT("CONNECTION", "Unable to resolve hostname '%s'", hostname);
    }

    ip_address = inet_ntoa(*((struct in_addr **)he->h_addr_list)[0]);

    LOG_INFO("CONNECTION", "Resolved IP-Address to '%s'", ip_address);

    // Set up the socket
    conn->remote_sockfd = socket(AF_INET, SOCK_STREAM, 0);

    if(conn->remote_sockfd == -1) {
        LOG_ERROR_AND_EXIT("CONNECTION", "Unable to open socket! Error: %s", strerror(errno));
    }

    setsockopt(conn->remote_sockfd, SOL_SOCKET, SO_REUSEADDR, NULL, 0);

    if(bind_address != NULL) {
        struct sockaddr_in local;

        local.sin_family = AF_INET;
        local.sin_addr.s_addr = inet_addr(bind_address);
        local.sin_port = 0;

        bind(conn->remote_sockfd, (struct sockaddr *) &local, sizeof(bind));
    }

    server.sin_family = AF_INET;
    server.sin_addr.s_addr = inet_addr(ip_address);
    server.sin_port = htons(port);

    // Connect to the server
    int ret = connect(conn->remote_sockfd, (struct sockaddr *) &server, sizeof(server));

    if(ret == -1) {
        LOG_ERROR_AND_EXIT("CONNECTION", "Unable to connect to server '%s'! Error: %s", hostname, strerror(errno));
    }

    LOG_INFO("CONNECTION", "Successfully established a TCP-connection to server '%s'!", hostname);

    // Exchange the infiniband connection information and set the queue to 'ready to send'
    __exchange_ib_connection_info(conn);

    set_qp_state_to_rtr(conn->queue_pair, conn->remote_conn_info.lid, conn->remote_conn_info.qpn);
    set_qp_state_to_rts(conn->queue_pair);
}

void connect_to_client(connection *conn, const char *bind_address, uint16_t port) {
    int sockfd;
    struct sockaddr_in server, client;

    LOG_INFO("CONNECTION", "Connecting to a client...");

    // Set up the socket
    sockfd = socket(AF_INET, SOCK_STREAM, 0);

    if(sockfd == -1) {
        LOG_ERROR_AND_EXIT("CONNECTION", "Unable to open socket! Error: %s", strerror(errno));
    }

    setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, NULL, 0);

    server.sin_family = AF_INET;
    server.sin_addr.s_addr = bind_address == NULL ? INADDR_ANY : inet_addr(bind_address);
    server.sin_port = htons(port);

    int ret = bind(sockfd, (struct sockaddr *)&server, sizeof(server));

    if(ret == -1) {
        close(sockfd);
        LOG_ERROR_AND_EXIT("CONNECTION", "Unable to bind socket! Error: %s", strerror(errno));
    }

    LOG_INFO("CONNECTION", "Waiting for an incoming connection...");

    // Wait for a client to connect
    listen(sockfd, 1);

    int addr_len = sizeof(struct sockaddr_in);
    conn->remote_sockfd = accept(sockfd, (struct sockaddr *)&client, (socklen_t *) &addr_len);

    if(conn->remote_sockfd == -1) {
        close(sockfd);
        LOG_ERROR_AND_EXIT("CONNECTION", "Error while accepting an incoming connection! Error: %s", strerror(errno));
    }

    LOG_INFO("CONNECTION", "Successfully established a TCP-connection to a client!")

    // Exchange the infiniband connection information and set the queue to 'ready to send'
    __exchange_ib_connection_info(conn);

    close(sockfd);

    set_qp_state_to_rtr(conn->queue_pair, conn->remote_conn_info.lid, conn->remote_conn_info.qpn);
    set_qp_state_to_rts(conn->queue_pair);
}

void close_connection(connection *conn) {
    LOG_INFO("CONNECTION", "Closing connection...");

    // Free all resources
    deregister_memory_region(conn->prot_dom, conn->send_mem_reg);
    deregister_memory_region(conn->prot_dom, conn->recv_mem_reg);

    free(conn->send_mem_reg->addr);
    free(conn->recv_mem_reg->addr);

    free(conn->send_mem_reg);
    free(conn->recv_mem_reg);

    free(conn->send_wrs);
    free(conn->recv_wrs);

    close_queue_pair(conn->queue_pair);
    free(conn->queue_pair);

    close(conn->remote_sockfd);

    conn->prot_dom = NULL;
    conn->send_comp_queue = NULL;
    conn->recv_comp_queue = NULL;
    conn->queue_pair = NULL;
    conn->send_mem_reg = NULL;
    conn->recv_mem_reg = NULL;
    conn->send_wrs = NULL;
    conn->recv_wrs = NULL;
    conn->remote_sockfd = 0;

    LOG_INFO("CONNECTION", "Successfully closed connection!");
}

void msg_send(connection *conn, uint32_t amount) {
    struct ibv_send_wr *bad_wr;

    if(amount == 0) {
        return;
    }

    for(uint32_t i = 0; i < amount; i++) {
        conn->send_wrs[i].wr_id = 0; // An id can be assigned to each work request. As we don't use it,
                                     // we just set it to 0.
        conn->send_wrs[i].sg_list = &conn->send_sge; // The scatter-gather element, that points to our send buffer
        conn->send_wrs[i].num_sge = 1; // We only have a single scatter-gather-element per work request
        conn->send_wrs[i].opcode = IBV_WR_SEND; // Opcode for sending messages
        conn->send_wrs[i].send_flags = IBV_SEND_SIGNALED; // Generate a work completion for every work request
        conn->send_wrs[i].next = i < amount - 1 ? &conn->send_wrs[i + 1] : NULL; // Chain work requests, so we can post
                                                                                 // all at once
    }

    // Post the work requests to the send queue
    int ret = ibv_post_send(conn->queue_pair->qp, conn->send_wrs, &bad_wr);

    if(ret != 0) {
        LOG_ERROR_AND_EXIT("CONNECTION", "Error while posting send work requests! Error: %s", strerror(ret));
    }
}

void msg_recv(connection *conn, uint32_t amount) {
    struct ibv_recv_wr *bad_wr;

    if(amount == 0) {
        return;
    }

    for(uint32_t i = 0; i < amount; i++) {
        conn->recv_wrs[i].wr_id = 0; // An id can be assigned to each work request. As we don't use it,
                                     // we just set it to 0.
        conn->recv_wrs[i].sg_list = &conn->recv_sge; // The scatter-gather element, that points to our receive buffer
        conn->recv_wrs[i].num_sge = 1; // We only have a single scatter-gather-element per work request
        conn->recv_wrs[i].next = i < amount - 1 ? &conn->recv_wrs[i + 1] : NULL; // Chain work requests, so we can post
                                                                                 // all at once
    }

    // Post the work requests to the receive queue
    int ret = ibv_post_recv(conn->queue_pair->qp, conn->recv_wrs, &bad_wr);

    if(ret != 0) {
        LOG_ERROR_AND_EXIT("CONNECTION", "Error while posting receive work requests! Error: %s", strerror(ret));
    }
}

void rdma_write(connection *conn, uint32_t amount) {
    struct ibv_send_wr *bad_wr;

    if(amount == 0) {
        return;
    }

    for(uint32_t i = 0; i < amount; i++) {
        conn->send_wrs[i].wr_id = 0; // An id can be assigned to each work request. As we don't use it,
                                     // we just set it to 0.
        conn->send_wrs[i].sg_list = &conn->send_sge; // The scatter-gather element, that points to our send buffer
        conn->send_wrs[i].num_sge = 1; // We only have a single scatter-gather-element per work request
        conn->send_wrs[i].opcode = IBV_WR_RDMA_WRITE; // Opcode for writing via RDMA
        conn->send_wrs[i].send_flags = IBV_SEND_SIGNALED; // Generate a work completion for every work request
        conn->send_wrs[i].wr.rdma.remote_addr = conn->remote_conn_info.remote_address; // Address of the memory region
                                                                                       // on the remote host
        conn->send_wrs[i].wr.rdma.rkey = conn->remote_conn_info.rkey; // Remote key of the remote memory region
        conn->send_wrs[i].next = i < amount - 1 ? &conn->send_wrs[i + 1] : NULL; // Chain work requests, so we can post
                                                                                 // all at once
    }

    // Post the work requests to the send queue
    int ret = ibv_post_send(conn->queue_pair->qp, conn->send_wrs, &bad_wr);

    if(ret != 0) {
        LOG_ERROR_AND_EXIT("CONNECTION", "Error while posting rdma write work requests! Error: %s", strerror(ret));
    }
}

void __exchange_ib_connection_info(connection *conn) {
    char msg[sizeof("0000:00000000:00000000:0000000000000000")];

    LOG_INFO("CONNECTION", "Exchanging infiniband connection info...");

    // Create a string, which contains the local ib connection info
    sprintf(msg, "%04hx:%08x:%08x:%016lx", conn->local_conn_info.lid, conn->local_conn_info.qpn,
            conn->local_conn_info.rkey, conn->local_conn_info.remote_address);

    // Write the info-string
    if(write(conn->remote_sockfd, msg, sizeof(msg)) != sizeof(msg)) {
        LOG_ERROR_AND_EXIT("CONNECTION", "Error while sending the infiniband connection info!");
    }

    // Read the remote connection's info-string
    if(read(conn->remote_sockfd, msg, sizeof(msg)) != sizeof(msg)) {
        LOG_ERROR_AND_EXIT("CONNECTION", "Error while receiving the remote infiniband connection info!");
    }

    // Parse the received information string
    if(sscanf(msg, "%hx:%x:%x:%lx", &conn->remote_conn_info.lid, &conn->remote_conn_info.qpn,
              &conn->remote_conn_info.rkey, &conn->remote_conn_info.remote_address) != 4) {
        LOG_ERROR_AND_EXIT("CONNECTION", "Error while parsing the received infiniband connection info!")
    }

    LOG_INFO("CONNECTION", "Successfully exchanged infiniband connection info! Received Lid: 0x%04x, Qpn: 0x%08x, "
                           "Rkey: 0x%08x, Remote address: 0x%016lx", conn->remote_conn_info.lid,
             conn->remote_conn_info.qpn, conn->remote_conn_info.rkey, conn->remote_conn_info.remote_address);
}

File: src/DisniBench/src/Connection.java
import com.ibm.disni.verbs.*;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.util.LinkedList;
import java.util.Objects;

/**
 * Connection to a remote host.
 *
 * This class allows connecting to either a server or a client and send messages to the remote host.
 *
 * @author Fabian Ruhland, HHU
 * @date 2019
 */
class Connection {

    /**
     * The send buffer.
     */
    private ByteBuffer sendBuf;

    /**
     * The receive buffer.
     */
    private ByteBuffer recvBuf;

    /**
     * The TCP-socket, that is used to exchange data with the remote host.
     */
    private Socket socket;

    /**
     * The jVerbs event channel.
     */
    private RdmaEventChannel eventChannel;

    /**
     * Parameters for a jVerbs connection.
     */
    private RdmaConnParam connectionParams;

    /**
     * The connection id.
     */
    private RdmaCmId id;

    /**
     * The size of the queue pair and completion queue.
     */
    private int queueSize;

    /**
     * The memory region, that is used for sending.
     */
    private IbvMr sendRegion;

    /**
     * The memory region, that is used for receiving.
     */
    private IbvMr recvRegion;

    /**
     * Wraps some of the JVerbs-functions and -objects.
     */
    private VerbsWrapper wrapper;

    /**
     * The key of the remote host's receive memory region.
     */
    private int remoteKey;

    /**
     * The address of the remote host's receive memory region.
     */
    private long remoteAddress;

    /**
     * List of scatter-gather elements used for sending.
     */
    private LinkedList<IbvSge> sendSges;

    /**
     * List of scatter-gather elements used for receiving.
     */
    private LinkedList<IbvSge> recvSges;

    /**
     * Reusable send work requests.
     */
    private IbvSendWR[] sendWrs;

    /**
     * Reusable receive work requests.
     */
    private IbvRecvWR[] recvWrs;

    /**
     * List of send work requests.
     */
    private LinkedList<IbvSendWR> sendWrList;

    /**
     * List of receive work requests.
     */
    private LinkedList<IbvRecvWR> recvWrList;

    /**
     * Create a connection.
     *
     * @param bufSize The size to be used for sendBuf and recvBuf
     * @param queueSize The queue size to be used for the InfiniBand queue pairs
     */
    Connection(int bufSize, int queueSize) {
        Log.INFO("CONNECTION", "Creating connection...");

        this.sendBuf = ByteBuffer.allocateDirect(bufSize);
        this.recvBuf = ByteBuffer.allocateDirect(bufSize);

        this.connectionParams = new RdmaConnParam();

        try {
            this.connectionParams.setInitiator_depth((byte) 1);
            this.connectionParams.setResponder_resources((byte) 1);
            this.connectionParams.setRetry_count((byte) 3);
            this.connectionParams.setRnr_retry_count((byte) 3);
        } catch (IOException e) {
            Log.WARN("CONNETION", "");
        }

        this.queueSize = queueSize;

        this.sendSges = new LinkedList<>();
        this.recvSges = new LinkedList<>();

        this.sendWrs = new IbvSendWR[queueSize];
        this.recvWrs = new IbvRecvWR[queueSize];

        for(int i = 0; i < this.sendWrs.length; i++) {
            this.sendWrs[i] = new IbvSendWR();
        }

        for(int i = 0; i < this.sendWrs.length; i++) {
            this.recvWrs[i] = new IbvRecvWR();
        }

        this.sendWrList = new LinkedList<>();
        this.recvWrList = new LinkedList<>();

        try {
            this.eventChannel = RdmaEventChannel.createEventChannel();
        } catch (IOException e) {
            Log.ERROR_AND_EXIT("Unable to create event channel! Error: %s", e.getMessage());
        }

        Log.INFO("CONNECTION", "Finished creating connection!");
    }

    /**
     * Connect to a remote server.
     *
     * @param bindAddress The address to bind the socket to (may be null, or empty string)
     * @param hostname The server's hostname
     * @param port The TCP-port
     */
    void connectToServer(String bindAddress, String hostname, int port) {
        /*Log.INFO("CONNECTION", "Connecting to server '%s'...", hostname);

        // Create connection id
        try {
            id = ConnectionId.create(eventChannel, PortSpace.RDMA_PS_TCP);
        } catch (Exception e) {
            Log.ERROR_AND_EXIT("Unable to create connection id! Error: %s", e.getMessage());
        }

        // Resolve address
        try {
            if (bindAddress == null || bindAddress.isEmpty()) {
                id.resolveAddress(new InetSocketAddress(port), new InetSocketAddress(hostname, port), 5000);
            } else {
                id.resolveAddress(new InetSocketAddress(bindAddress, port),
                        new InetSocketAddress(hostname, port), 5000);
            }

            ConnectionEvent event = eventChannel.getConnectionEvent(-1);

            if(event.getEventType() != ConnectionEvent.EventType.RDMA_CM_EVENT_ADDR_RESOLVED) {
                Log.ERROR_AND_EXIT("CONNECTION",
                        "Unable to resolve address! Error: Received wrong event type '%s'", event.getEventType());
            }

            eventChannel.ackConnectionEvent(event);
        } catch (Exception e) {
            Log.ERROR_AND_EXIT("CONNECTION", "Unable to resolve address! Error: %s", e.getMessage());
        }

        Log.INFO("CONNECTION", "Successfully resolved address!");

        // Resolve route
        try {
            id.resolveRoute(5000);

            ConnectionEvent event = eventChannel.getConnectionEvent(-1);

            if(event.getEventType() != ConnectionEvent.EventType.RDMA_CM_EVENT_ROUTE_RESOLVED) {
                Log.ERROR_AND_EXIT("CONNECTION",
                        "Unable to resolve address! Error: Received wrong event type '%s'", event.getEventType());
            }

            eventChannel.ackConnectionEvent(event);
        } catch (Exception e) {
            Log.ERROR_AND_EXIT("CONNECTION", "Unable to resolve route! Error: %s", e.getMessage());
        }

        Log.INFO("CONNECTION", "Successfully resolved route!");

        // Establish connection
        try {
            wrapper = new JVerbsWrapper(id, queueSize);

            sendRegion = wrapper.registerMemoryRegion(sendBuf);
            recvRegion = wrapper.registerMemoryRegion(recvBuf);

            ScatterGatherElement sendSge = new ScatterGatherElement();
            sendSge.setAddress(sendRegion.getAddress());
            sendSge.setLength(sendRegion.getLength());
            sendSge.setLocalKey(sendRegion.getLocalKey());
            sendSges.add(sendSge);

            ScatterGatherElement recvSge = new ScatterGatherElement();
            recvSge.setAddress(recvRegion.getAddress());
            recvSge.setLength(recvRegion.getLength());
            recvSge.setLocalKey(recvRegion.getLocalKey());
            recvSges.add(recvSge);

            id.connect(connectionParams);

            ConnectionEvent event = eventChannel.getConnectionEvent(-1);

            if(event.getEventType() != ConnectionEvent.EventType.RDMA_CM_EVENT_ESTABLISHED) {
                Log.ERROR_AND_EXIT("CONNECTION",
                        "Unable to connect to remote host! Error: Received wrong event type '%s'",
                        event.getEventType());
            }

            eventChannel.ackConnectionEvent(event);
        } catch (Exception e) {
            Log.ERROR_AND_EXIT("CONNECTION", "Unable to connect to remote host! Error: %s", e.getMessage());
        }

        try {
            Thread.sleep(1000);
            socket = new Socket();

            if(bindAddress == null || bindAddress.isEmpty()) {
                socket.bind(new InetSocketAddress(port));
            } else {
                socket.bind(new InetSocketAddress(bindAddress, port));
            }

            socket.connect(new InetSocketAddress(hostname, port));
        } catch (Exception e) {
            Log.ERROR_AND_EXIT("CONNECTION", "Unable to open socket! Error: %s", e.getMessage());
        }

        exchangeRdmaInfo();

        Log.INFO("CONNECTION", "Successfully connected to remote host!");*/
    }

    /**
     * Connect to a remote client.
     *
     * @param bindAddress The address to bind the socket to (may be null, or empty string)
     * @param port The TCP-port to listen on
     */
    void connectToClient(String bindAddress, int port) {
        Log.INFO("CONNECTION", "Connecting to a client...");

        RdmaCmId serverId = null;
        ServerSocket serverSocket = null;

        try {
            serverId = eventChannel.createId(RdmaCm.RDMA_PS_TCP);
        } catch (Exception e) {
            Log.ERROR_AND_EXIT("Unable to create connection id! Error: %s", e.getMessage());
        }

        try {
            if (bindAddress == null || bindAddress.isEmpty()) {
                Objects.requireNonNull(serverId).bindAddr(new InetSocketAddress(port));
            } else {
                Objects.requireNonNull(serverId).bindAddr(new InetSocketAddress(bindAddress, port));
            }
        } catch (IOException | NullPointerException e) {
            Log.ERROR_AND_EXIT("CONNECTION", "Unable to bind address! Error: %s", e.getMessage());
        }

        Log.INFO("CONNECTION", "Waiting for an incoming connection...");

        try {
            Objects.requireNonNull(serverId).listen(0);

            RdmaCmEvent event = eventChannel.getCmEvent(-1);

            if(event.getEvent() != RdmaCmEvent.EventType.RDMA_CM_EVENT_CONNECT_REQUEST.ordinal()) {
                Log.ERROR_AND_EXIT("CONNECTION",
                        "Error while accepting an incoming connection! Error: Received wrong event type '%d'",
                        event.getEvent());
            }

            event.ackEvent();

            id = event.getConnIdPriv();
        } catch (Exception e) {
            Log.ERROR_AND_EXIT("CONNECTION", "Error while accepting an incoming connection! Error: %s",
                    e.getMessage());
        }

        Log.INFO("CONNECTION", "Received connection request!");

        try {
            wrapper = new VerbsWrapper(id, queueSize);

            sendRegion = wrapper.registerMemoryRegion(sendBuf);
            recvRegion = wrapper.registerMemoryRegion(recvBuf);

            IbvSge sendSge = new IbvSge();
            sendSge.setAddr(sendRegion.getAddr());
            sendSge.setLength(sendRegion.getLength());
            sendSge.setLkey(sendRegion.getLkey());
            sendSges.add(sendSge);

            IbvSge recvSge = new IbvSge();
            recvSge.setAddr(recvRegion.getAddr());
            recvSge.setLength(recvRegion.getLength());
            recvSge.setLkey(recvRegion.getLkey());
            recvSges.add(recvSge);

            id.accept(connectionParams);

            RdmaCmEvent event = eventChannel.getCmEvent(-1);

            if (event.getEvent() != RdmaCmEvent.EventType.RDMA_CM_EVENT_ESTABLISHED.ordinal()) {
                Log.ERROR_AND_EXIT("CONNECTION",
                        "Error while accepting an incoming connection! Error: Received wrong event type '%d'",
                        event.getEvent());
            }

            event.ackEvent();
        } catch(Exception e) {
            Log.ERROR_AND_EXIT("CONNECTION",
                    "Error while accepting an incoming connection! Error: Received wrong event type '%s'",
                    e.getMessage());
        }

        try {
            serverSocket = new ServerSocket();

            if(bindAddress == null || bindAddress.isEmpty()) {
                serverSocket.bind(new InetSocketAddress(port));
            } else {
                serverSocket.bind(new InetSocketAddress(bindAddress, port));
            }

            socket = Objects.requireNonNull(serverSocket).accept();
        } catch (Exception e) {
            Log.ERROR_AND_EXIT("CONNECTION", "Error while accepting an incoming connection! Error: %s",
                    e.getMessage());
        }

        exchangeRdmaInfo();

        Log.INFO("CONNECTION", "Successfully established a connection to a client!");
    }

    /**
     * Exchange the address and remote key of the receive memory regions with the remote host, so that rdma can be used.
     */
    private void exchangeRdmaInfo() {
        try {
            DataInputStream inputStream = new DataInputStream(socket.getInputStream());
            DataOutputStream outputStream = new DataOutputStream(socket.getOutputStream());

            byte[] outputBuf = String.format("%08x:%016x", recvRegion.getRkey(), recvRegion.getAddr()).getBytes();
            byte[] inputBuf = new byte[outputBuf.length];

            outputStream.write(outputBuf);
            inputStream.readFully(inputBuf);

            String[] data = new String(inputBuf).split(":");

            remoteKey = Integer.parseUnsignedInt(data[0], 16);
            remoteAddress = Long.parseUnsignedLong(data[1], 16);

            Log.INFO("CONNECTION", "Successfully exchanged rdma information with the remote host! " +
                    "Received: 0x%08x, 0x%016x", remoteKey, remoteAddress);
        } catch(Exception e) {
            Log.ERROR_AND_EXIT("CONNECTION",
                    "Unable to exchange rdma information with the remote host! Error: %s", e.getMessage());
        }
    }

    /**
     * Disconnect from the remote host.
     */
    void close() {
        Log.INFO("CONNECTION", "Closing connection...");

        try {
            socket.close();

            id.disconnect();

            RdmaCmEvent event = eventChannel.getCmEvent(-1);

            if (event.getEvent() != RdmaCmEvent.EventType.RDMA_CM_EVENT_DISCONNECTED.ordinal()) {
                Log.WARN("CONNECTION", "Error while disconnecting! Error: Received wrong event type '%d'",
                        event.getEvent());
            }

            event.ackEvent();

            wrapper.destroy();

            id.disconnect();
            id.destroyId();

            eventChannel.destroyEventChannel();
        } catch (Exception e) {
            Log.WARN("CONNECTION", "Unable to close connection! Error: %s", e.getMessage());

            return;
        }

        Log.INFO("CONNECTION", "Successfully closed connection!");
    }

    /**
     * Send a specified amount of messages to the remote host.
     *
     * @param msgCount The amount of messages to be sent
     *
     * @throws Exception DataOutputStream.write() may throw an Exception
     */
    void sendMessages(long msgCount) throws IOException {
        if(msgCount <= 0) {
            return;
        }

        sendWrList.clear();

        for(int i = 0; i < msgCount; i++) {
            sendWrs[i].setWr_id(1);
            sendWrs[i].setSg_list(sendSges);
            sendWrs[i].setOpcode(IbvSendWR.IBV_WR_SEND);
            sendWrs[i].setSend_flags(IbvSendWR.IBV_SEND_SIGNALED);

            sendWrList.add(sendWrs[i]);
        }

        SVCPostSend sendMethod = wrapper.getPostSendMethod(sendWrList);

        sendMethod.execute();
    }

    /**
     * Receive a specified amount of messages from the remote host.
     *
     * @param msgCount The amount of messages to be sent
     */
    void recvMessages(long msgCount) throws Exception {
        if(msgCount <= 0) {
            return;
        }

        recvWrList.clear();

        for(int i = 0; i < msgCount; i++) {
            recvWrs[i].setWr_id(1);
            recvWrs[i].setSg_list(recvSges);

            recvWrList.add(recvWrs[i]);
        }

        SVCPostRecv receiveMethod = wrapper.getPostReceiveMethod(recvWrList);

        receiveMethod.execute();
    }

    /**
     * Use RDMA to write a given amount of times to the remote host.
     *
     * @param count The amount of rdma writes to be performed
     */
    void rdmaWrite(long count) throws IOException {
        if(count <= 0) {
            return;
        }

        sendWrList.clear();

        for(int i = 0; i < count; i++) {
            sendWrs[i].setWr_id(1);
            sendWrs[i].setSg_list(sendSges);
            sendWrs[i].setOpcode(IbvSendWR.IBV_WR_RDMA_WRITE);
            sendWrs[i].setSend_flags(IbvSendWR.IBV_SEND_SIGNALED);
            sendWrs[i].getRdma().setRemote_addr(remoteAddress);
            sendWrs[i].getRdma().setRkey(remoteKey);

            sendWrList.add(sendWrs[i]);
        }

        SVCPostSend sendMethod = wrapper.getPostSendMethod(sendWrList);

        sendMethod.execute();
    }

    /**
     * Use RDMA to read a given amount of times from the remote host.
     *
     * @param count The amount of rdma reads to be performed
     */
    void rdmaRead(long count) throws IOException {
        if(count <= 0) {
            return;
        }

        sendWrList.clear();

        for(int i = 0; i < count; i++) {
            sendWrs[i].setWr_id(1);
            sendWrs[i].setSg_list(sendSges);
            sendWrs[i].setOpcode(IbvSendWR.IBV_WR_RDMA_READ);
            sendWrs[i].setSend_flags(IbvSendWR.IBV_SEND_SIGNALED);
            sendWrs[i].getRdma().setRemote_addr(remoteAddress);
            sendWrs[i].getRdma().setRkey(remoteKey);

            sendWrList.add(sendWrs[i]);
        }

        SVCPostSend sendMethod = wrapper.getPostSendMethod(sendWrList);

        sendMethod.execute();
    }

    /**
     * Poll completions from the completion queue.
     *
     * @return The amount of polled work completions
     */
    int pollCompletionQueue(VerbsWrapper.CqType type) throws IOException {
        SVCPollCq pollMethod = wrapper.getPollCqMethod(type);

        pollMethod.execute();

        int polled = pollMethod.getPolls();

        IbvWC[] workComps = wrapper.getWorkCompletions();

        for(int i = 0; i < polled; i++) {
            if(workComps[i].getStatus() != IbvWC.IbvWcStatus.IBV_WC_SUCCESS.ordinal()) {
                Log.ERROR_AND_EXIT("CONNECTION", "Work completion failed! Status: %d" +
                        workComps[i].getStatus());
            }
        }

        return polled;
    }

    /**
     * Get the queue size of the InfiniBand queue pairs.
     */
    int getQueueSize() {
        return queueSize;
    }

    /**
     * Get the socket, that is used to exchange the rdma parameters.
     */
    Socket getSocket() {
        return socket;
    }
}
